engine/access/rest/websockets/controller.go

```diff
@@ -129,23 +132,31 @@ func (c *Controller) configureKeepalive() error {
 // Parameters:
 // - ctx: Context to control cancellation and timeouts.
 func (c *Controller) monitorInactivity(ctx context.Context) error {
+	defer c.inactivityTracker.Stop()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
-		default:
+		case <-c.inactivityTracker.C():
 			if c.dataProviders.Size() == 0 {
-				<-time.After(c.config.InactivityTimeout)
-
-				if c.dataProviders.Size() == 0 {
-					// TODO: decide if we need to write response message with reason to client
-					return fmt.Errorf("no recent activity for %v", c.config.InactivityTimeout)
-				}
+				// Optionally send a message to the client indicating the reason for closure.
+				c.logger.Info().Msg("Connection inactive, closing due to timeout.")
+				return fmt.Errorf("no recent activity for %v", c.config.InactivityTimeout)
 			}
 		}
 	}
 }
 
+// checkInactivity checks if there are no active data providers
+// and resets the inactivity timer. This function should be called after removing a data provider
+// to update the inactivity tracker and ensure it reflects the current state.
+func (c *Controller) checkInactivity() {
+	if c.dataProviders.Size() == 0 {
+		c.inactivityTracker.Reset(c.config.InactivityTimeout)
```

Comment:
according to the docs, this may not be safe:
```
// For a Timer created with NewTimer, Reset should be invoked only on
// stopped or expired timers with drained channels.
```

this code doesn't guarantee that the timer is drained when calling reset here.